lib/psql.js

```diff
@@ -24,32 +17,136 @@ function env (db) {
   })
 }
 
+function tunnelConfig (db) {
+  const localHost = '127.0.0.1'
+  const localPort = Math.floor(Math.random() * (65535 - 49152) + 49152)
+  return {
+    username: 'bastion',
+    host: db.bastionHost,
+    privateKey: db.bastionKey,
+    dstHost: db.host,
+    dstPort: db.port,
+    localHost: localHost,
+    localPort: localPort
+  }
+}
+
+function sshTunnel (db, dbTunnelConfig, timeout) {
+  return new Promise((resolve, reject) => {
+    // if necessary to tunnel, setup a tunnel
+    // see also https://github.com/heroku/heroku/blob/master/lib/heroku/helpers/heroku_postgresql.rb#L53-L80
+    if (db.bastionKey) {
+      tunnel(dbTunnelConfig, (err, tnl) => {
+        if (err) {
+          debug(err)
+          reject(`Unable to establish a secure tunnel to your database.`)
+        }
+        debug('Tunnel created')
+        resolve(tnl)
+      })
+    } else {
+      resolve()
+    }
+    setTimeout(() => reject('Establishing a secure tunnel timed out'), timeout)
+  })
+}
+
+function execPsql (query, dbEnv, timeout) {
+  const {spawn} = require('child_process')
+  const stripEOF = require('strip-eof')
+  return new Promise((resolve, reject) => {
+    let result = ''
+    let psql = spawn('psql', ['-c', query], {env: dbEnv, encoding: 'utf8'})
+    psql.stdout.on('data', function (data) {
+      result += data.toString()
+    })
+    psql.on('close', function (code) {
+      resolve(result)
+    })
+    psql.stderr.on('data', (data) => {
+      debug(stripEOF(data.toString('utf8')))
+      reject(`Unable to execute the command.`)
+      process.exit(1)
+    })
+    psql.on('error', (err) => {
+      if (err.code === 'ENOENT') {
+        reject(`The local psql command could not be located. For help installing psql, see https://devcenter.heroku.com/articles/heroku-postgresql#local-setup`)
+      } else {
+        reject(err)
+      }
+    })
+    setTimeout(() => reject('psql call timed out'), timeout)
+  })
+}
+
+function psqlInteractive (dbEnv, prompt, timeout) {
+  const {spawn} = require('child_process')
+  return new Promise((resolve, reject) => {
+    let psql = spawn('psql',
+                     ['--set', `PROMPT1=${prompt}`, '--set', `PROMPT2=${prompt}`],
+                     {env: dbEnv, stdio: 'inherit'})
+    psql.on('error', (err) => {
+      if (err.code === 'ENOENT') {
+        reject(`The local psql command could not be located. For help installing psql, see https://devcenter.heroku.com/articles/heroku-postgresql#local-setup`)
+      } else {
+        reject(err)
+      }
+    })
+    resolve()
```

Comment:
Don't you want to wait for the program to exit (`.on('close')`) to resolve this?